### PR TITLE
Feature/bitmex

### PR DESCRIPTION
**Pull request by notVitaliy**
@gijsmouha This looks great!

Can you explain why there needs to be an abstraction for leveraged/margin versus cash accounts? Is this a bitmex specific implementation?

### CODE REVIEW

**Comment by Eyelessian** [comment]
@notVitaliy , going to reply to you this way since i can't comment on your comment for some reason :cry: 
To elaborate on your question, the abstraction is not Bitmex specific, it's a feature used for leverage trading and there are several exchanges that use this type of trading. 
Basically, what the original bot supported (afaik, i can be mistaken) was trading from one type of currency to another. Leverage trading is a whole different system. If you want to try it, you can use the [Bitmex testnet](https://testnet.bitmex.com) which allows users to try it out. The bot can work with that same testnet as well btw :smile: 
I have to say that the code is not what i want it to be yet, the bot works with my changes but i feel that much things can be improved further. So if you want to merge it into your repo, feel free to do so but i advise that you do it in a separate branch instead of master. I haven't had the time to work on it further :sob: 
I hope that this made my code a little bit more clear to you, if not. Feel free to contact me :smile:

File: src/engine/order.ts [hunk 1]
```diff
@@ -52,12 +52,13 @@ export class OrderEngine {
...
   public async executeBuy(quote?: number, strength = 1) {
     const { exchange, symbol } = this.opts
+    logger.verbose(`Executing buy order on exchange ${exchange}, symbol ${symbol}`)
+
     const rawPrice = quote ? quote : await this.quoteEngine.getBuyPrice()
     const price = this.exchangeProvider.priceToPrecision(exchange, symbol, rawPrice)
 
     const purchasingPower = this.exchangeProvider.priceToPrecision(exchange, symbol, this.wallet.currency * strength)
     const amount = this.exchangeProvider.amountToPrecision((purchasingPower / price) * 0.995)
-
```
Comment (sgmheyhey) [inline src/engine/order.ts:60]:
```
error TS2554: Expected 3 arguments, but got 1.

61     const amount = this.exchangeProvider.amountToPrecision((purchasingPower / price) * 0.995)
                                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: src/exchange/exchange.provider.ts [hunk 2]
```diff
@@ -94,8 +111,8 @@ export class ExchangeProvider {
     return this.exchanges.get(exchange).getTradeCursor(trade)
   }
 
-  public amountToPrecision(amount: number) {
-    return Math.floor(amount * 100000000) / 100000000
+  public amountToPrecision(exchange: string, amount: number, currentPrice: number) {
+    return this.exchanges.get(exchange).amountToPrecision(amount, currentPrice)
   }
```
Comment (sgmheyhey) [inline src/exchange/exchange.provider.ts:116]:
```
  src/exchange/exchange.provider.ts:114:46
    114   public amountToPrecision(exchange: string, amount: number, currentPrice: number) {
                                                     ~~~~~~~~~~~~~~
    An argument for 'amount' was not provided.
```

File: src/exchange/exchangeTypes/currencyExchange.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    public amountToPrecision(amount: number) {
+        return Math.floor(amount * 100000000) / 100000000
+    }
+
+    public async createOrder(symbol: string, type: string, side: string, amount: number, price: number): Promise<Order> {
+        const fn = () => this.exchangeConnection.createOrder(symbol, type, side, amount, price)
+        return this.bottleneck.schedule(fn)
+      }
```
Comment (sgmheyhey) [inline src/exchange/exchangeTypes/currencyExchange.ts:23]:
```
src/exchange/exchange.wrapper.ts:36:88 - error TS2345: Argument of type '{ [key: string]: number | Date; }' is not assignable to parameter of type 'Params'.
  Index signatures are incompatible.
    Type 'number | Date' is not assignable to type 'string | number'.
      Type 'Date' is not assignable to type 'string | number'.
        Type 'Date' is not assignable to type 'number'.

36     const fn = () => this.exchangeConnection.fetchTrades(symbol, undefined, undefined, params)
                                                                                          ~~~~~~

src/exchange/exchangeTypes/currencyExchange.ts:21:70 - error TS2345: Argument of type 'string' is not assignable to parameter of type '"market" | "limit"'.

21         const fn = () => this.exchangeConnection.createOrder(symbol, type, side, amount, price)  
```

File: src/exchange/adapters/base.ts [hunk 4]
```diff
@@ -40,7 +40,7 @@ export interface ExchangeAdapter {
   scan: 'back' | 'forward'
   ratelimit: number
   getTradeCursor: (trade: Trade) => number
-  mapTradeParams: (start: number) => { [key: string]: number }
+  mapTradeParams: (start: number) => { [key: string]: number| Date }
 }
```
Comment (sgmheyhey) [inline src/exchange/adapters/base.ts:44]:
```
src/store/adjustment.store.ts:21:36 - error TS2345: Argument of type '{ currency: number; asset: number; type: "user" | "cancelOrder" | "newOrder" | "fillOrder"; timestamp: number; exchange: string; symbol: string; strategy?: string; sessionId: string; }' is not assignable to parameter of type 'AdjustmentCollection'.
  Property 'time' is missing in type '{ currency: number; asset: number; type: "user" | "cancelOrder" | "newOrder" | "fillOrder"; timestamp: number; exchange: string; symbol: string; strategy?: string; sessionId: string; }' but required in type '{ sessionId: string; exchange: string; symbol: string; strategy: string; time: number; }'.

21     await dbDriver.adjustment.save(data)
                                      ~~~~

  src/lib/db/db.types.ts:120:3
    120   time: number
          ~~~~
    'time' is declared here.
```

**Pull request closed without merging**